Firestore/core/src/api/load_bundle_task.h

```diff
@@ -0,0 +1,197 @@
+ * limitations under the License.
+ */
+#ifndef FIRESTORE_CORE_SRC_API_LOAD_BUNDLE_TASK_H_
+#define FIRESTORE_CORE_SRC_API_LOAD_BUNDLE_TASK_H_
+
+#include <array>
+#include <cstdint>
+#include <memory>
+#include <mutex>  // NOLINT(build/c++11)
+#include <string>
+#include <vector>
+
+#include "Firestore/core/src/util/executor.h"
+
+namespace firebase {
+namespace firestore {
+namespace api {
+
+/**
+ * Represents the state of bundle loading tasks.
+ *
+ * Both `Success` and `Error` are final states: task will abort
+ * or complete and there will be no more updates after they are reported.
+ */
+enum class LoadBundleTaskState { Error = 0, InProgress = 1, Success = 2 };
+
+/** Represents a progress update or a final state from loading bundles. */
+class LoadBundleTaskProgress {
+ public:
+  LoadBundleTaskProgress() {
+  }
+  LoadBundleTaskProgress(uint32_t documents_loaded,
+                         uint32_t total_documents,
+                         uint64_t bytes_loaded,
+                         uint64_t total_bytes,
+                         LoadBundleTaskState state)
+      : documents_loaded_(documents_loaded),
+        total_documents_(total_documents),
+        bytes_loaded_(bytes_loaded),
+        total_bytes_(total_bytes),
+        state_(state) {
+  }
+
+  /** Returns how many documents have been loaded. */
+  uint32_t documents_loaded() const {
+    return documents_loaded_;
+  }
+
+  /**
+   * Returns the total number of documents in the bundle. Returns 0 if the
+   * bundle failed to parse.
+   */
+  uint32_t total_documents() const {
+    return total_documents_;
+  }
+
+  /** Returns how many bytes have been loaded. */
+  uint64_t bytes_loaded() const {
+    return bytes_loaded_;
+  }
+
+  /** Returns the total number of bytes in the bundle. Returns 0 if the bundle
+   * failed to parse. */
+  uint64_t total_bytes() const {
+    return total_bytes_;
+  }
+
+  /** Returns the current state of the task. */
+  LoadBundleTaskState state() const {
+    return state_;
+  }
+
+  void set_state(LoadBundleTaskState state) {
+    state_ = state;
+  }
+
+ private:
+  uint32_t documents_loaded_ = 0;
+  uint32_t total_documents_ = 0;
+  uint64_t bytes_loaded_ = 0;
+  uint64_t total_bytes_ = 0;
+
+  LoadBundleTaskState state_ = LoadBundleTaskState::InProgress;
+};
+
+inline bool operator==(const LoadBundleTaskProgress lhs,
+                       const LoadBundleTaskProgress& rhs) {
+  return lhs.state() == rhs.state() &&
+         lhs.bytes_loaded() == rhs.bytes_loaded() &&
+         lhs.documents_loaded() == rhs.documents_loaded() &&
+         lhs.total_bytes() == rhs.total_bytes() &&
+         lhs.total_documents() == rhs.total_documents();
+}
+
+inline bool operator!=(const LoadBundleTaskProgress lhs,
+                       const LoadBundleTaskProgress& rhs) {
+  return !(lhs == rhs);
+}
+
+/** A handle used to lookup and remove observer from the task. */
+using LoadBundleHandle = std::string;
```

Comment:
It seems a little overkill. Passing strings around is also less efficient (higher memory consumption, less efficient comparisons), though it's unlikely to matter much. Since you can simply monotonically increase the integer, it doesn't seem significantly more complicated than using an UUID. I think it would be a simplification, unless I'm missing something.

Reply:
Done.